Use internode mTLS for ringpop

folbricht-stripe opened the pull request:
<!-- Describe what has changed in this PR -->
**What changed?**

Internode mTLS now applies to the ringpop traffic as well.

<!-- Tell your future self why have you made these changes -->
**Why?**

The ringpop traffic was previously not encrypted, even when Temporal was configured for internode mTLS.


<!-- How have you verified this change? Tested locally? Added a unit test? Checked in staging env? -->
**How did you test it?**

- Tested locally with a small 2-node cluster in docker-compose. Confirmed admin commands worked and showed both nodes in the ring.
- PCAP'ed the traffic between nodes to confirm it's encrypted

I have not deployed/tested this in a large cluster yet.

<!-- Assuming the worst case, what can be broken when deploying this change to production? -->
**Potential risks**

If this breaks, the cluster likely won't form as nodes can't talk to each other over ringpop.


<!-- Is this PR a hotfix candidate or require that a notification be sent to the broader community? (Yes/No) -->
**Is hotfix candidate?**

No

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/temporalio/temporal?pullRequest=1902) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/temporalio/temporal?pullRequest=1902) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/temporalio/temporal?pullRequest=1902) it.</sub>

wxing1292 commented on common/rpc/rpc.go:
```diff
@@ -151,15 +151,36 @@ func (d *RPCFactory) GetRingpopChannel() *tchannel.Channel {
+			d.logger.Fatal("Failed to get internode TLS server config", tag.Error(err))
+		}
+		if serverTLSConfig != nil {
+			l, err := tls.Listen("tcp", ringpopHostAddress, serverTLSConfig)
```
`l` -> can we use a longer var name? at least 3 chars

folbricht-stripe replied:
Done, calling it `tlsListener` now which is more descriptive

nagl-stripe commented:
What will happen during an online upgrade when this change hits a cluster of temporal instances?

paulnpdev commented:
maxim just asked the same question on Slack


On Wed, Sep 8, 2021 at 5:21 PM nagl-stripe ***@***.***> wrote:

> What will happen during an online upgrade when this change hits a cluster
> of temporal instances?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/temporalio/temporal/pull/1902#issuecomment-915660904>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AUMR7H5SOEL723D6JXKQTTDUA74ZTANCNFSM5DV2AG4Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

nagl-stripe commented:
> maxim just asked the same question on Slack

I'm genuinely curious. I assume it will cause a split brain scenario, but I don't know what the actual impact of that will be.

samarabbas reviewed (changes requested):
Thanks @folbricht-stripe for this awesome contribution.  I think we first need to upgrade tchannel-go to latest dependency as the version we are using is v1.22.0 which unfortunately does not have support for tls.  This support was added much later in this [PR](https://github.com/uber/tchannel-go/commit/74b0fff2ec2047a53774898f8b23abf021ed5f7a).  I think we first need to upgrade ringpop-go to pick a more recent release of tchannel-go which has this support.  Until then tls dialer you are passing through options is going to be ignored.

samarabbas reviewed (changes requested):
Can you also include both positive and negative test for enabling tls for ringpop?

alexshtin commented:
I believe it might work. Ringpop actually depends on interfaces and if I understand the code correctly the `tchannel.Channel` instance from 1.22 (which is actually latest and only 23 days old) should work fine there.

Would love to see tests for this.

alexshtin commented on common/rpc/rpc.go:
```diff
@@ -151,15 +151,36 @@ func (d *RPCFactory) GetRingpopChannel() *tchannel.Channel {
+		if err != nil {
+			d.logger.Fatal("Failed to get internode TLS client config", tag.Error(err))
+		}
+		opts := new(tchannel.ChannelOptions)
```
```suggestion
		opts := &tchannel.ChannelOptions{}
```

alexshtin commented on common/rpc/rpc.go:
```diff
@@ -151,15 +151,36 @@ func (d *RPCFactory) GetRingpopChannel() *tchannel.Channel {
+			}
+
+			if err := d.ringpopChannel.Serve(l); err != nil {
+				d.logger.Fatal("Failed to start ringpop listener", tag.Error(err), tag.Address(ringpopHostAddress))
```
```suggestion
				d.logger.Fatal("Failed to serve ringpop listener", tag.Error(err), tag.Address(ringpopHostAddress))
```

samarabbas commented:
> I believe it might work. Ringpop actually depends on interfaces and if I understand the code correctly the `tchannel.Channel` instance from 1.22 (which is actually latest and only 23 days old) should work fine there.
> 
> Would love to see tests for this.

@alexshtin is right.  I got confused by glide.yaml files in ringpop-go repo which is pointing to a very old version (1.8) of tchannel-go.  Looks like we are currently on 1.16 which has the custom dialer changes so this should work.

folbricht-stripe commented:
I added some tests to make sure the listener (and clients) used for ringpop do use TLS when configured. Might need a couple more cases, but wanted to check if that's what you had in mind wrt testing first.

alexshtin commented on common/rpc/test/rpc_localstore_tls_test.go:
```diff
@@ -689,3 +733,46 @@ func (s *localStoreRPCSuite) validateTLSInfo(tlsInfo *credentials.TLSInfo, err e
+	runRingpopTLSTest(s.Suite, s.logger, s.insecureRPCFactory, s.ringpopServerTLSRPCFactoryB, false)
+}
+
+func runRingpopTLSTest(s suite.Suite, logger log.Logger, serverA *TestFactory, serverB *TestFactory, isValid bool) {
```
I think it would be better if:
```suggestion
func runRingpopTLSTest(s suite.Suite, logger log.Logger, serverA *TestFactory, serverB *TestFactory, expectError bool) {
```
Also flip values on caller side and `if` checks bellow.

alexshtin reviewed (commented):
I like the approach. If you have few more cases in mind, please add them and I believe we are good to go.

alexshtin commented:
@folbricht-stripe, any updates? Can you fix minor comments left and I will merge this PR.

folbricht-stripe commented:
Sorry, I'm traveling so a bit slow to respond. I addressed the comments and squashed it now. Thinking about the tests a bit longer, I think they're sufficient.

The only open question is what would happen if an existing cluster is updated with mTLS support. It's likely going to lead to a split-brain situation during the update I imagine but not sure what the impact would be.

alexshtin commented:
Disable this by default with ability to enable with dynamic config: #1967.

samarabbas commented:
@folbricht-stripe thanks for this amazing contribution to Temporal.
